src/array.jl

```diff
@@ -769,29 +769,30 @@ function levels!(A::CategoricalArray{T, N, R}, newlevels::Vector;
     end
     (levels(A) == newlevels) && return A # nothing to do
 
-    # prepare a map from new levels to their refs
-    newlv2ref = Dict{eltype(newlevels), Tuple{Int, Int}}()
+    # map each new level to its ref code and number of occurrences
+    newlv2ref = Dict{eltype(newlevels), Pair{Int, Int}}()
     dupnewlvs = similar(newlevels, 0)
     for (i, lv) in enumerate(newlevels)
-        lvref = get(newlv2ref, lv, (i, 0))
-        newlv2ref[lv] = (lvref[1], lvref[2] + 1)
+        lvref = get(newlv2ref, lv, i => 0)
```

Comment:
Why not just `if occursin(newlv2ref, lv); push!(dupnewlvs, lv); continue; end`? To avoid the double dict lookup (when getting and then setting), you could even do `if get!(newlv2ref, lv, i) != i`.

Reply:
I wanted to avoid `unique(dupnewlvs)`. But probably your suggestion is less expensive on average and the code is simpler.